File: src/application/config/loader.ts
import { constants } from "node:fs";
import { homedir } from "node:os";
import { access, readFile, readdir, stat } from "node:fs/promises";
import { basename, extname, join, resolve } from "node:path";
import { parse as parseYaml } from "yaml";
import { DEFAULT_PROJECT_PLAIN } from "./defaults.js";
import { configSchema } from "./schema.js";
import type { LoadedProjectConfig, ProjectConfig } from "./types.js";
import { validateConfigReferences } from "./validation.js";
import { isPlainRecord, mergeYamlLayers } from "./yaml-merge.js";

export async function safeStat(pathLike: string) {
  try {
    return await stat(pathLike);
  } catch {
    return undefined;
  }
}

export function parseYamlTagged(pathLike: string, raw: string): unknown {
  try {
    return parseYaml(raw) as unknown;
  } catch (error: unknown) {
    const suffix = error instanceof Error ? error.message : String(error);
    throw new Error(`${pathLike}: ${suffix}`);
  }
}

export async function findDefaultConfigPath(): Promise<string | undefined> {
  const candidate = resolve("rlm.config.yaml");
  try {
    await access(candidate, constants.R_OK);
    return candidate;
  } catch {
    return undefined;
  }
}

export async function loadScopedFragments(scopeRoot: string): Promise<Record<string, unknown>> {
  let accumulator: Record<string, unknown> = {};
  const scopeStat = await safeStat(scopeRoot);
  if (!scopeStat?.isDirectory()) {
    return {};
  }

  const cfgPath = join(scopeRoot, "config.yaml");
  const cfgStat = await safeStat(cfgPath);
  if (cfgStat?.isFile()) {
    const yamlRoot = parseYamlTagged(cfgPath, await readFile(cfgPath, "utf8"));
    if (yamlRoot === null || yamlRoot === undefined || !isPlainRecord(yamlRoot)) {
      throw new Error(`${cfgPath}: expected YAML mapping at root`);
    }

    accumulator = mergeYamlLayers(accumulator, yamlRoot);
  }

  const agentsDir = join(scopeRoot, "agents");
  const agentsStat = await safeStat(agentsDir);
  if (agentsStat?.isDirectory()) {
    const entries = [...(await readdir(agentsDir))].sort((aName, bName) =>
      aName.localeCompare(bName),
    );
    const agentsPartial: Record<string, unknown> = {};
    for (const fileName of entries) {
      if (!fileName.endsWith(".yaml") && !fileName.endsWith(".yml")) continue;
      const agentId = basename(fileName, extname(fileName)).trim();
      if (!agentId) continue;

      const filePath = join(agentsDir, fileName);
      const agentDoc = parseYamlTagged(filePath, await readFile(filePath, "utf8"));
      if (!isPlainRecord(agentDoc)) {
        throw new Error(`${filePath}: agent fragment must map tools/models`);
      }

      agentsPartial[agentId] = agentDoc;
    }

    if (Object.keys(agentsPartial).length > 0) {
      accumulator = mergeYamlLayers(accumulator, { agents: agentsPartial });
    }
  }

  const modelsDir = join(scopeRoot, "models");
  const modelsStat = await safeStat(modelsDir);
  if (modelsStat?.isDirectory()) {
    const tiers: Record<string, unknown> = {};
    const entries = [...(await readdir(modelsDir))].sort((aName, bName) =>
      aName.localeCompare(bName),
    );
    for (const entryName of entries) {
      if (!entryName.endsWith(".yaml") && !entryName.endsWith(".yml")) continue;
      const tierKey = basename(entryName, extname(entryName)).trim();
      if (!tierKey) continue;

      const filePath = join(modelsDir, entryName);
      const tierDoc = parseYamlTagged(filePath, await readFile(filePath, "utf8"));
      if (!isPlainRecord(tierDoc)) {
        throw new Error(`${filePath}: tier fragment must map fields`);
      }

      tiers[tierKey] = tierDoc;
    }

    if (Object.keys(tiers).length > 0) {
      accumulator = mergeYamlLayers(accumulator, { models: { tiers } });
    }
  }

  return accumulator;
}

export async function loadProjectConfig(path?: string): Promise<LoadedProjectConfig> {
  if (path) {
    const resolvedExplicit = resolve(path);
    const raw = await readFile(resolvedExplicit, "utf8");
    const yamlRoot = parseYamlTagged(resolvedExplicit, raw);
    let mergedYaml = structuredClone(DEFAULT_PROJECT_PLAIN);
    mergedYaml = mergeYamlLayers(
      mergedYaml,
      yamlRoot !== null && isPlainRecord(yamlRoot) ? yamlRoot : {},
    );

    const config = configSchema.parse(mergedYaml) as ProjectConfig;
    validateConfigReferences(config);

    return {
      config,
      path: resolvedExplicit,
    };
  }

  let mergedYaml = structuredClone(DEFAULT_PROJECT_PLAIN);

  const globalFragments = join(homedir(), ".rlm");
  if ((await safeStat(globalFragments))?.isDirectory()) {
    mergedYaml = mergeYamlLayers(mergedYaml, await loadScopedFragments(globalFragments));
  }

  let primaryPath: string | undefined;

  const cwd = process.cwd();
  const scopedDir = join(cwd, ".rlm");

  const legacyScoped = resolve(cwd, "rlm.config.yaml");
  const legacyExists = (await safeStat(legacyScoped))?.isFile() ?? false;

  if ((await safeStat(scopedDir))?.isDirectory()) {
    if (legacyExists) {
      const legacyYamlRaw = await readFile(legacyScoped, "utf8");
      const legacyYaml = parseYamlTagged(legacyScoped, legacyYamlRaw);
      if (!legacyYaml || !isPlainRecord(legacyYaml)) {
        throw new Error(`${legacyScoped}: expected YAML mapping at root`);
      }

      mergedYaml = mergeYamlLayers(mergedYaml, legacyYaml);
      primaryPath ??= legacyScoped;
    }

    mergedYaml = mergeYamlLayers(mergedYaml, await loadScopedFragments(scopedDir));

    const cfgChild = join(scopedDir, "config.yaml");
    if ((await safeStat(cfgChild))?.isFile()) {
      primaryPath ??= cfgChild;
    }

    primaryPath ??= legacyExists ? legacyScoped : undefined;
  } else {
    const discoveredPath = legacyExists ? legacyScoped : await findDefaultConfigPath();
    if (discoveredPath) {
      const discoveredRaw = await readFile(discoveredPath, "utf8");
      const parsedDoc = parseYamlTagged(discoveredPath, discoveredRaw);
      if (!isPlainRecord(parsedDoc)) {
        throw new Error(`${discoveredPath}: expected YAML mapping at root`);
      }

      mergedYaml = mergeYamlLayers(mergedYaml, parsedDoc);
      primaryPath ??= discoveredPath;
    }
  }

  const config = configSchema.parse(mergedYaml) as ProjectConfig;
  validateConfigReferences(config);

  return {
    config,
    path: primaryPath,
  };
}

File: crates/rlm-core/src/application/config/loader.rs
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use serde_json::{json, Map, Value};

use crate::persistence::ProjectPaths;

use super::defaults::default_project_plain;
use super::validation::{validate_config_references, validate_config_shape};
use super::yaml_merge::{is_plain_record, merge_yaml_layers};

#[derive(Debug, Clone)]
pub struct LoadedProjectConfig {
    pub config: Value,
    pub path: Option<PathBuf>,
}

pub fn load_project_config(
    project_root: &Path,
    explicit_path: Option<&Path>,
) -> io::Result<LoadedProjectConfig> {
    if let Some(path) = explicit_path {
        return load_explicit(path);
    }

    let paths = ProjectPaths::from_root(project_root.to_path_buf());
    let mut merged = default_project_plain();

    let global_fragments = dirs::home_dir()
        .map(|home| home.join(".rlm"))
        .unwrap_or_else(|| PathBuf::from(".rlm"));
    if global_fragments.is_dir() {
        merged = merge_yaml_layers(merged, load_scoped_fragments(&global_fragments)?);
    }

    let mut primary_path: Option<PathBuf> = None;
    let legacy_scoped = paths.legacy_config_path();
    let legacy_exists = legacy_scoped.is_file();

    if paths.scoped_dir().is_dir() {
        if legacy_exists {
            let legacy_yaml = parse_yaml_file(&legacy_scoped)?;
            merged = merge_yaml_layers(merged, legacy_yaml);
            primary_path.get_or_insert(legacy_scoped.clone());
        }
        merged = merge_yaml_layers(merged, load_scoped_fragments(&paths.scoped_dir())?);
        let cfg_child = paths.scoped_config_path();
        if cfg_child.is_file() {
            primary_path.get_or_insert(cfg_child);
        }
        if primary_path.is_none() && legacy_exists {
            primary_path = Some(legacy_scoped);
        }
    } else if legacy_exists {
        let parsed = parse_yaml_file(&legacy_scoped)?;
        merged = merge_yaml_layers(merged, parsed);
        primary_path = Some(legacy_scoped);
    } else if let Some(discovered) = find_default_config_path(project_root) {
        let parsed = parse_yaml_file(&discovered)?;
        merged = merge_yaml_layers(merged, parsed);
        primary_path = Some(discovered);
    }

    validate_config_shape(&merged)?;
    validate_config_references(&merged)?;
    Ok(LoadedProjectConfig {
        config: merged,
        path: primary_path,
    })
}

fn load_explicit(path: &Path) -> io::Result<LoadedProjectConfig> {
    let resolved = path.canonicalize().unwrap_or_else(|_| path.to_path_buf());
    let yaml_root = parse_yaml_file(&resolved)?;
    let merged = merge_yaml_layers(default_project_plain(), yaml_root);
    validate_config_shape(&merged)?;
    validate_config_references(&merged)?;
    Ok(LoadedProjectConfig {
        config: merged,
        path: Some(resolved),
    })
}

fn find_default_config_path(project_root: &Path) -> Option<PathBuf> {
    let candidate = project_root.join("rlm.config.yaml");
    candidate.is_file().then_some(candidate)
}

pub(crate) fn parse_yaml_file(path: &Path) -> io::Result<Value> {
    let raw = fs::read_to_string(path)
        .map_err(|err| io::Error::new(err.kind(), format!("{}: {err}", path.display())))?;
    serde_yaml::from_str(&raw).map_err(|err| {
        io::Error::new(
            io::ErrorKind::InvalidData,
            format!("{}: {err}", path.display()),
        )
    })
}

fn load_scoped_fragments(scope_root: &Path) -> io::Result<Value> {
    if !scope_root.is_dir() {
        return Ok(json!({}));
    }

    let mut accumulator = Map::new();
    let cfg_path = scope_root.join("config.yaml");
    if cfg_path.is_file() {
        let yaml_root = parse_yaml_file(&cfg_path)?;
        if !is_plain_record(&yaml_root) {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("{}: expected YAML mapping at root", cfg_path.display()),
            ));
        }
        accumulator = merge_yaml_layers(Value::Object(accumulator), yaml_root)
            .as_object()
            .cloned()
            .unwrap_or_default();
    }

    let agents_dir = scope_root.join("agents");
    if agents_dir.is_dir() {
        let mut agents_partial = Map::new();
        let mut entries: Vec<_> = fs::read_dir(&agents_dir)?
            .flatten()
            .map(|entry| entry.file_name().to_string_lossy().to_string())
            .collect();
        entries.sort();
        for file_name in entries {
            if !(file_name.ends_with(".yaml") || file_name.ends_with(".yml")) {
                continue;
            }
            let agent_id = file_name
                .trim_end_matches(".yaml")
                .trim_end_matches(".yml")
                .trim();
            if agent_id.is_empty() {
                continue;
            }
            let file_path = agents_dir.join(&file_name);
            let agent_doc = parse_yaml_file(&file_path)?;
            if !is_plain_record(&agent_doc) {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!(
                        "{}: agent fragment must map tools/models",
                        file_path.display()
                    ),
                ));
            }
            agents_partial.insert(agent_id.to_string(), agent_doc);
        }
        if !agents_partial.is_empty() {
            accumulator = merge_yaml_layers(
                Value::Object(accumulator),
                json!({ "agents": agents_partial }),
            )
            .as_object()
            .cloned()
            .unwrap_or_default();
        }
    }

    let models_dir = scope_root.join("models");
    if models_dir.is_dir() {
        let mut tiers = Map::new();
        let mut entries: Vec<_> = fs::read_dir(&models_dir)?
            .flatten()
            .map(|entry| entry.file_name().to_string_lossy().to_string())
            .collect();
        entries.sort();
        for entry_name in entries {
            if !(entry_name.ends_with(".yaml") || entry_name.ends_with(".yml")) {
                continue;
            }
            let tier_key = entry_name
                .trim_end_matches(".yaml")
                .trim_end_matches(".yml")
                .trim();
            if tier_key.is_empty() {
                continue;
            }
            let file_path = models_dir.join(&entry_name);
            let tier_doc = parse_yaml_file(&file_path)?;
            if !is_plain_record(&tier_doc) {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!("{}: tier fragment must map fields", file_path.display()),
                ));
            }
            tiers.insert(tier_key.to_string(), tier_doc);
        }
        if !tiers.is_empty() {
            accumulator = merge_yaml_layers(
                Value::Object(accumulator),
                json!({ "models": { "tiers": tiers } }),
            )
            .as_object()
            .cloned()
            .unwrap_or_default();
        }
    }

    Ok(Value::Object(accumulator))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn parse_yaml_includes_path_context() {
        let dir = std::env::temp_dir().join(format!("rlm-config-{}", std::process::id()));
        let _ = fs::remove_dir_all(&dir);
        fs::create_dir_all(&dir).unwrap();
        let path = dir.join("bad.yaml");
        fs::write(&path, ":\n- not a map\n").unwrap();
        let err = parse_yaml_file(&path).unwrap_err();
        assert!(err.to_string().contains(&path.display().to_string()));
    }
}
